Add a financial aid approval module

900miles opened the pull request:
Adds a new program module that allows program admins to easily view and approve financial aid applications in bulk. 
Looks something like this:
![image](https://user-images.githubusercontent.com/2653008/46183785-6bd08000-c287-11e8-9506-a79a36d68338.png)

Fixes #2485

Also updates the program module docs to add a how-to section and an example module. It could use some more work, but something is better than nothing!

(I don't quite know what happened with program_modules.rst, but I don't know how to fix it, sorry. I did add a new section to it though, so I can't just drop the file entirely.)

willgearty commented on esp/templates/program/modules/finaidapprovemodule/finaid.html:
```diff
@@ -0,0 +1,71 @@
+    <div id="program_form">
+      <table align="center">
+        <tr>
+          <th><input type="checkbox" onClick="toggle(this)" /></th>
```
Very nice addition!

willgearty commented on esp/esp/program/modules/handlers/finaidapprovemodule.py:
```diff
@@ -0,0 +1,120 @@
+                if str(req.user.id) not in userchecklist:
+                    continue
+
+                if is_blank(req.household_income) and is_blank(req.extra_explaination):
```
Hmm...this made sense back when it was a script, but maybe this should be toggled by an option in the template, now? I think this is fine as the default, but I could see cases where people might just want to approve all of them, regardless of whether these fields have information.

willgearty commented on esp/esp/program/modules/handlers/finaidapprovemodule.py:
```diff
@@ -0,0 +1,120 @@
+                return x is None #or re.match(r'^(\s)*$', x)
+
+            for req in reqs:
+                if str(req.user.id) not in userchecklist:
```
Since only this page will be posting to this page, is there any reason this would ever be true?

benjaminjkraft replied:
If someone requested finaid between the time you loaded the page and the time you submitted!

willgearty commented on esp/esp/program/modules/handlers/finaidapprovemodule.py:
```diff
@@ -0,0 +1,120 @@
+            # ITERATE & APPROVE REQUESTS
+            userchecklist = request.POST.getlist("user")
+
+            message += "{} New Approval{} ({} error{}): "
```
This is a lot of string manipulation and concatenation. Might make more sense to just maintain two lists of users (approved and errors) and then loop over them in the template.

willgearty commented on esp/esp/program/modules/handlers/finaidapprovemodule.py:
```diff
@@ -0,0 +1,120 @@
+
+
+        # Remove trailing commas or colons (if no new approvals) from the message
+        if (message[-2:] == ", " or message[-2:] == ": "):
```
If you use lists, this will be much easier to do in the template.

willgearty commented on esp/templates/program/modules/finaidapprovemodule/finaid.html:
```diff
@@ -0,0 +1,71 @@
+          <th><input type="checkbox" onClick="toggle(this)" /></th>
+          <th>Approved?</th>
+          <th width="70px">Name</th>
+          <th width="70px">Household Income</th>
```
Can you also include the `reduced_lunch` and `student_prepare` attributes as well?

900miles replied:
Added! What should I label the student_prepare column in the template?

willgearty replied:
Well, the help text for this field is "Did anyone besides the student fill out any portions of this form?" (True then means that someone besides the student prepared the answers), which seems to be the opposite of the field name (which implies that True means that the student prepared the answers). I was thinking something like "Student Prepared" or "Prepared by Student" would be a good label for the column, but then you'd want to show the inverse of the actual field value (I think). If you don't want that, you could use "Non-Student Prepared" or "Prepared by non-Student" or something like that, I guess?

900miles replied:
I went with "Filled Out Only by Student". Does that sound good? I don't want some unsuspecting admin to be confused by what it's trying to portray. :)

benjaminjkraft commented on docs/dev/program_modules.rst:
```diff
@@ -15,7 +15,84 @@ Program modules are an ESP website-specific framework for bits of code associate
+
+Program modules live in python files in ``/esp/esp/program/modules/handlers/``. (Note that other utilities mentioned in this section live in the same file.)
+
+Any HTML web pages included in a module live in ``/esp/templates/program/modules/``
```
```suggestion
Any HTML web pages included in a module live in ``/esp/templates/program/modules/``.
```

benjaminjkraft commented on docs/dev/program_modules.rst:
```diff
@@ -15,7 +15,84 @@ Program modules are an ESP website-specific framework for bits of code associate
+
+For a list of the program modules, and what they do, see ``/docs/admin/program_modules.rst``.
+
+How To Create A New Module
```
Thanks for adding this!

benjaminjkraft commented on esp/esp/program/modules/handlers/finaidapprovemodule.py:
```diff
@@ -0,0 +1,120 @@
+        if len(reqs) == 0:
+            message = "No requests found."
+            context["message"] = message
+            return render_to_response(self.baseDir()+'finaid.html', request, context)
```
Eh, I think it's ok to do it this way: it makes it much clearer which parts are relevant when, even if we end up duplicating the `render_to_response` call.

bhorowitz commented:
This looks like an awesome new module! I bet programs will love it! :)

willgearty commented:
Tested this on the Stanford site, and it works well, but I also noticed that it approves applications with blank details (not sure how this happens, but it does), so you'll want to make sure `is_blank()` checks for `""` in addition to `None`.

willgearty commented on esp/templates/program/modules/finaidapprovemodule/finaid.html:
```diff
@@ -0,0 +1,71 @@
+{% block xtrajs %}
+<script type="text/javascript" language="javascript">
+function toggle(source) {
+  checkboxes = document.getElementsByName('user');
```
You should be able to do all of this with a single line of jQuery:
`$j('input[name=user]').prop('checked', source.checked);`

willgearty commented:
I'd say this fixes #811.

willgearty commented:
@900miles, have you had a chance to make the requested/suggested changes?

900miles commented:
@willgearty, sorry I haven't responded in a while. I've made most of the changes, just a couple left to finish up!

willgearty commented:
Sounds good! Looking forward to seeing to the changes!

900miles commented:
@willgearty All the changes have been made. Let me know if there's anything else you want me to do!

willgearty commented:
This pull request **introduces 3 alerts** when merging d9e4a7a23db99371f1049fabf585188d0c075432 into 3ed2822e482f7c92c2c442e8ab182c26ce0b3510 - [view on LGTM.com](https://lgtm.com/projects/g/learning-unlimited/ESP-Website/rev/pr-dee8bdb62a4d9da80d5eaeebc03ee57bf5f8536a)

**new alerts:**

* 2 for Unused import
* 1 for Except block handles &#39;BaseException&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

willgearty commented on esp/esp/program/modules/handlers/finaidapprovemodule.py:
```diff
@@ -59,62 +59,51 @@ class Meta:
 
             def is_blank(x):
-                return x is None #or re.match(r'^(\s)*$', x)
+                return x is None or x == ""
```
I think you can catch both of these cases with `bool(x)`, so you might not even need a separate function.

900miles replied:
I tried it out and ``bool(x)`` does work, although I'm thinking the separate function makes it easier to tell what the program is doing:
``if is_blank(req.household_income) or is_blank(req.extra_explaination):``
vs.
``if bool(req.household_income) and bool(req.extra_explaination):``
Although maybe not.. What do you think?

willgearty replied:
I agree it's clearer with the function, and it's probably just as efficient time-wise.

willgearty commented on esp/esp/program/modules/handlers/finaidapprovemodule.py:
```diff
@@ -0,0 +1,109 @@
+                    users_error.append(req.user.name())
+
+        context["requests"] = reqs
+        context["users_approved"] = users_approved
```
Looks like you also need to add `users_error` to `context`

willgearty commented on esp/templates/program/modules/finaidapprovemodule/finaid.html:
```diff
@@ -0,0 +1,104 @@
+  {% endif %}
+
+  {% if users_error %}
+  <h3 style="padding-top: 10px; color: firebrick"> {{ users_approved|length }} Errors</h3>
```
```suggestion
  <h3 style="padding-top: 10px; color: firebrick"> {{ users_error|length }} Errors</h3>
```

willgearty commented on esp/templates/program/modules/finaidapprovemodule/finaid.html:
```diff
@@ -0,0 +1,104 @@
+          <td align="center">${{ req.household_income }}</td>
+          <td>{{ req.extra_explaination }}</td>
+          <td align="center">
+            {% if req.student_prepare == False %}
```
```suggestion
            {% if not req.student_prepare %}
```

willgearty commented:
This pull request **introduces 3 alerts** when merging c8db2f2bd4b4bba6428b58f62c33294b87006012 into 764f5c7b387d43b8d11f5d1061caa9c8b7dbdb23 - [view on LGTM.com](https://lgtm.com/projects/g/learning-unlimited/ESP-Website/rev/pr-81475174f51d13c3d38a0f94b426e3f5a91f893c)

**new alerts:**

* 2 for Unused import
* 1 for Except block handles &#39;BaseException&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

900miles commented:
> This pull request **introduces 3 alerts** when merging [c8db2f2](https://github.com/learning-unlimited/ESP-Website/commit/c8db2f2bd4b4bba6428b58f62c33294b87006012) into [764f5c7](https://github.com/learning-unlimited/ESP-Website/commit/764f5c7b387d43b8d11f5d1061caa9c8b7dbdb23) - [view on LGTM.com](https://lgtm.com/projects/g/learning-unlimited/ESP-Website/rev/pr-81475174f51d13c3d38a0f94b426e3f5a91f893c)
> 
> **new alerts:**
> 
>     * 2 for Unused import
> 
>     * 1 for Except block handles 'BaseException'
> 
> 
> _Comment posted by [LGTM.com](https://lgtm.com)_

Fixed the unused imports and other changes. Not sure what to do about the "except block" one though?

willgearty commented:
This pull request **introduces 3 alerts** when merging 7b8e76597edc7c2a26b67363ca1fc4f4dcce98ea into 764f5c7b387d43b8d11f5d1061caa9c8b7dbdb23 - [view on LGTM.com](https://lgtm.com/projects/g/learning-unlimited/ESP-Website/rev/pr-69fd93a7e10a6f9da8313076134c32f7528fca78)

**new alerts:**

* 2 for Unused import
* 1 for Except block handles &#39;BaseException&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

900miles commented:
@willgearty Whoops! I pushed them now. 
Also, I found that sometimes when approving new requests, the entire user list is reshuffled on the page. If that's a problem, perhaps it would be better to sort the list before populating the table..

willgearty commented:
This pull request **introduces 1 alert** when merging fb6e436d708c17c2e0015ad8223d7174feac11c7 into 1e4fa5bbcc67e85097105c61aeb1e0179f502f7c - [view on LGTM.com](https://lgtm.com/projects/g/learning-unlimited/ESP-Website/rev/pr-518bb74d267ead91274d0b6965980d0b8bddd12f)

**new alerts:**

* 1 for Except block handles &#39;BaseException&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

willgearty commented:
@900miles I hadn't realized they weren't in any order, but yeah, it would probably make the most sense to sort the financial aid requests by the name of the student. Once that's done, this looks good to me!

900miles commented:
@willgearty I decided to order the requests by most recent instead of last name, as I feel that could be more useful to admin. And with that, I should be done!

willgearty commented:
This pull request **introduces 1 alert** when merging a1c000d2ddc0b455734c649ed9bc431513a4d8d4 into 1e4fa5bbcc67e85097105c61aeb1e0179f502f7c - [view on LGTM.com](https://lgtm.com/projects/g/learning-unlimited/ESP-Website/rev/pr-2f077dc9e3cc96ef8dda4f37d53730d9252b949c)

**new alerts:**

* 1 for Except block handles &#39;BaseException&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

willgearty commented:
Now that I think about it, that's probably better if you end up approving financial aid multiple times. The module works perfectly, and I think the logic is definitely done. Unfortunately, it looks like the table doesn't display that well in thinner themes, like fruitsalad:

![image](https://user-images.githubusercontent.com/7232514/54793569-430ac100-4c00-11e9-97d8-b409e7b545af.png)

or circles:
![image](https://user-images.githubusercontent.com/7232514/54793659-ad236600-4c00-11e9-9e1b-7aee7df7f0e9.png)

It looks like the culprit is due to you hardcoding the width of `#program_form` to 800px, which is actually wider than the content area of most themes (outside of big picture). Perhaps a better option would be to set the width to 100% or max-width to 800px or something like that. In fact, just getting rid of this single CSS style makes the table look pretty good in circles:
![image](https://user-images.githubusercontent.com/7232514/54793757-4488b900-4c01-11e9-8311-7738825c6f78.png)
Tables can also often be finicky and override widths if they think it's for the best, so it's probably best to also convert to percentages instead of pixels for the columns widths, and maybe include something like `table-layout:fixed`.

Whatever you decide on, it'd be best to check how it looks in multiple themes.

P.S. Sorry for the long review process, but since you're making an entire new module with python logic, HTML, and CSS, that's often the case.

900miles commented:
@willgearty 
Okay, I changed ``width=800px`` to ``max-width=800px``. And instead of changing the column widths to percents instead of pixels, I changed them to ``5.3em``. Only the "Name" and "Income" columns have this, because I feel they need a certain width in order for the names/incomes to read properly, regardless of table size. ``em``s will scale with font size but not table width, so I feel they will be a better fit. The other columns have no set width.
Also, I added a couple of lines to override some formatting issues with the "floaty" theme.

Thank you for all your help!!

willgearty commented:
Looks good on all of the themes I tried! Great job, and thanks for all of the hard work on this! Looking forward to seeing what you work on next @900miles!
